Update Volume API

mauriciopoppe opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
/kind feature

**What this PR does / why we need it**:

Updates the Volume v1beta3 API according to the API review in #130, in addition to that:

- Regenerated the pb.go file for the changes I made in the proto file and `client/groups.volume/v1beta3/client_generated.go`, I synced these changes back to the vendor folder too so that the server can use them
- Updated the host wrapper in the server to use the new signatures
- I moved the host interface `API` to `internal/os/volume/api.go` because it belongs here, previously the owner was the wrapper, this fixed an issue where the host API wasn't fulfilling any interface and generated problems while renaming things
- Updated `integrationtests/volume_test.go` to use `diskNumber` (uint32) instead of diskID (string)
- ~~Regenerated the `internal/server/volume/internal/v1X` files, **the older APIs are incompatible with the new parameters in the v1beta3 proto file**, I didn't want to touch the autogenerated for older APIs but unfortunately I can't compile the code if I don't make these changes~~
- Regenerated the `internal/server/volume/internal/v1X` files, for pre v1beta3 versions I added more conversion fields to `conversion.go`, that way the internal server representation of the requests from any API is the same (no special cases for pre v1beta3 APIs aside from the pre v1beta3 methods)
- I updated the tests, there was one test that was checking an invalid case with an empty diskID but because we're using `diskNumber` I removed this test and instead I'm checking for negative values

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #130

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Updated the v1beta3 Volume API because of an API review, changes are summarized in #130 
```

mauriciopoppe commented:
/uncc @kkmsft @andyzhangx 
/cc @ddebroy @msau42 @jingxu97

mauriciopoppe commented on integrationtests/disk_test.go:
```diff
@@ -74,40 +74,41 @@ func TestDiskAPIGroup(t *testing.T) {
 		require.NoError(t, err, "failed setting disk offline, out=%v", out)
 
-		getReq := &v1beta3.GetAttachStateRequest{DiskID: diskNum}
+		getReq := &v1beta3.GetAttachStateRequest{DiskID: diskNumAsString}
```
this change is temporal, once we work in the DiskAPI and it accepts numbers I'll remove this

jingxu97 replied:
maybe create an issue to track this?

mauriciopoppe replied:
this was fixed in #140

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -10,27 +10,52 @@ service Volume {
-    // IsVolumeFormatted checks if a volume is formatted with NTFS
+    // UnmountVolume gracefully unmounts a volume
+    rpc UnmountVolume(UnmountVolumeRequest) returns (UnmountVolumeResponse) {}
+    // IsVolumeFormatted checks if a volume is formatted.
```
add . to all comments?

mauriciopoppe replied:
fixed.

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -10,27 +10,52 @@ service Volume {
     rpc FormatVolume(FormatVolumeRequest) returns (FormatVolumeResponse) {}
     // ResizeVolume performs resizing of the partition and file system for a block based volume
     rpc ResizeVolume(ResizeVolumeRequest) returns (ResizeVolumeResponse) {}
+    // GetVolumeStats gathers DiskSize, VolumeSize and VolumeUsedSize for a volume
```
update comments "gathers total types and used bytes for a volume?"

mauriciopoppe replied:
fixed

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -10,27 +10,52 @@ service Volume {
-    // Disk device ID of the disk to query for volumes
-    string disk_id = 1;
+    // Disk device number of the disk to query for volumes
+    int64 disk_number = 1;
```
i think int is enough

mauriciopoppe replied:
I agree that `int32` is enough however I'd like to keep everything as `int64` because all of the other int like fields in this file are `int64`

mauriciopoppe replied:
I take that back, I'm going to change it to `uint32` as that's the field type returned by `GetDisk`

jingxu97 commented on internal/server/volume/internal/v1beta1/conversion.go:
```diff
@@ -1,39 +1,39 @@
-	out.Mount = in.Mount
-	return nil
-}
+// // Add manual conversion functions here to override automatic conversion functions
```
so all of these are not needed?

mauriciopoppe replied:
These were removed in v1beta2, in any case each release of csi-proxy is tied to its own API version so this file isn't used anyways. i.e. if we release v1beta3 you can only connect to it using a v1beta3 client and the server API used would be the v1beta3 API

mauriciopoppe replied:
I removed these commented lines, probably @ddebroy can help confirm if these are actually needed in v1beta3

mauriciopoppe replied:
these lines were needed! After I understood how the conversion functions work I made more changes here and luckily was able to test most of the endpoints in the integration tests

mauriciopoppe commented:
I need to change disk number to `uint32`
/hold

mauriciopoppe commented:
/unhold

ddebroy commented on internal/server/volume/internal/v1beta1/conversion_generated.go:
```diff
@@ -10,8 +10,6 @@ import (
 )
 
 func autoConvert_v1beta1_DismountVolumeRequest_To_internal_DismountVolumeRequest(in *v1beta1.DismountVolumeRequest, out *internal.DismountVolumeRequest) error {
-	out.VolumeId = in.VolumeId
```
Wouldn't these removals in this file cause problems for the beta1 Volume API group?

mauriciopoppe replied:
you're correct, after reviewing the comments in the slack thread https://kubernetes.slack.com/archives/CN5JCCW31/p1621011691055500 I understood that the server side should be backwards compatible, I made changes in the server implementation and recreated the _generated.go files, I saw no changes in the generated files for pre v1beta3 APIs which is great!

ddebroy commented on internal/server/volume/server.go:
```diff
@@ -69,37 +49,41 @@ func (s *Server) MountVolume(context context.Context, request *internal.MountVol
-	klog.V(5).Infof("DismountVolume: Request: %+v", request)
-	response := &internal.DismountVolumeResponse{}
+func (s *Server) DismountVolume(context.Context, *internal.DismountVolumeRequest, apiversion.Version) (*internal.DismountVolumeResponse, error) {
+	return nil, fmt.Errorf("DismountVolume is deprecated, use UnmountVolume")
```
This is the first time we are making the Deprecation announcement. We should not block an existing API from working. Would a DismountVolume from a previous API version get routed to UnmountVolume?

mauriciopoppe replied:
got it, yes, I'm routing it to `UnmountVolume` instead

ddebroy commented on internal/server/volume/server.go:
```diff
@@ -181,6 +165,10 @@ func (s *Server) ResizeVolume(context context.Context, request *internal.ResizeV
 }
 
 func (s *Server) VolumeStats(context context.Context, request *internal.VolumeStatsRequest, version apiversion.Version) (*internal.VolumeStatsResponse, error) {
+	return nil, fmt.Errorf("DismountVolume is deprecated, use UnmountVolume")
```
The error message is incorrect here. Also, as stated above, we want to make sure a plugin invoking VolumeStats should be able to get the desired functionality and not end up with any error.

mauriciopoppe replied:
thanks for catching this, I think I removed the `deprecated` messages and instead all the previous requests are routed to the internal volume implementation

ddebroy reviewed (commented):
This is great! A few comments. I think for the deprecated APIs, we need to make sure all the deprecated APIs from previous API versions continue to work.

mauriciopoppe commented:
/hold

There are pending changes for backwards compatibility

mauriciopoppe commented:
@ddebroy I implemented some of the changes we talked about in our meeting yesterday:

- The v1beta3 volume client code no longer includes old methods or old parameters
- I've added conversion functions in v1alpha1, v1beta1 and v1beta2 from the client requests to the internal server representation of the requests
- I've removed all the copying of parameters in the server API handlers based on the request version (instead this is handled by the conversion functions)

Key files to look at:

- client/api/volume/v1beta3/api.proto
- client/groups/volume/v1beta3/client_generated.go 
- internal/server/volume/internal/types.go

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -5,128 +5,129 @@ package v1beta3;
-    // DismountVolume gracefully dismounts a volume
-    rpc DismountVolume(DismountVolumeRequest) returns (DismountVolumeResponse) {}
-    // IsVolumeFormatted checks if a volume is formatted with NTFS
+    // UnmountVolume gracefully unmounts a volume.
```
UnmountVolume flushes data cache to disk and removes the global staging path?

mauriciopoppe replied:
fixed

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -5,128 +5,129 @@ package v1beta3;
-    // GetVolumeIDFromMount gets the volume id for a given mount
-    rpc GetVolumeIDFromMount(VolumeIDFromMountRequest) returns (VolumeIDFromMountResponse) {}
-    // WriteVolumeCache write volume cache to disk
+    // GetVolumeStats gathers total types and used bytes for a volume.
```
total bytes

mauriciopoppe replied:
O: fixed!

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -5,128 +5,129 @@ package v1beta3;
+    rpc GetVolumeStats(GetVolumeStatsRequest) returns (GetVolumeStatsResponse) {}
+    // GetDiskNumberFromVolumeID gets the disk number of the disk where the volume is located.
+    rpc GetDiskNumberFromVolumeID(GetDiskNumberFromVolumeIDRequest) returns (GetDiskNumberFromVolumeIDResponse ) {}
+    // GetVolumeIDFromTargetPath gets the volume id for a given mount.
```
for a given target path?

mauriciopoppe replied:
ack

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -5,128 +5,129 @@ package v1beta3;
+    // Disk device number of the disk to query for volumes.
+    uint32 disk_number = 1;
+    // The partition number.
+    uint32 partition_number = 2;
```
mention partition number is optional. By default, it uses the first partition on the disk

mauriciopoppe replied:
proto3 doesn't have a way to mark this field as optional, I've added the comment that it's optional though

ddebroy reviewed (approved):
structure LGTM modulo other comments from Jing.

mauriciopoppe commented on internal/server/volume/internal/v1alpha1/conversion_generated.go:
```diff
@@ -128,25 +128,13 @@ func Convert_internal_IsVolumeFormattedResponse_To_v1alpha1_IsVolumeFormattedRes
 }
 
 func autoConvert_v1alpha1_ListVolumesOnDiskRequest_To_internal_ListVolumesOnDiskRequest(in *v1alpha1.ListVolumesOnDiskRequest, out *internal.ListVolumesOnDiskRequest) error {
```
right, it looks like the autoConvert functions will always be generated, I think that making changes to these _generated.go files is error prone because at some point we might delete them and want to auto generate them (at least that's the workflow that I usually follow) and the manual changes will be gone.

The easiest approach is to just ignore the fact that they're unused, we could document that they're unused and that would be it.

Another approach would be to modify the generator to stop generating these `auto` functions when it detects that there's a conversion function for it.

mauriciopoppe commented on integrationtests/volume_test.go:
```diff
@@ -196,14 +205,39 @@ func negativeVolumeTests(t *testing.T) {
 
+// clientCompatibilityTests tests that the API is compatible with versions that are before
+// the latest, e.g. that a v1beta2 client can still use the server csi-proxy v1beta3
+func clientCompatibilityTests(t *testing.T) {
```
No problem, thanks for suggesting the approach to add conversion functions, even though it added duplicated code in the conversion functions now the internal server structs no longer have info about client details.

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -5,128 +5,129 @@ package v1beta3;
 
 message ResizeVolumeRequest {
-    // Volume device ID of the volume to dismount
+    // Volume device ID of the volume to dismount.
```
dismount -- > resize

mauriciopoppe replied:
fixed

jingxu97 commented on integrationtests/volume_test.go:
```diff
@@ -212,11 +246,6 @@ func negativeDiskTests(t *testing.T) {
-	// Empty disk Id test
-	runNegativeListVolumeRequest(t, client, "")
-	// Negative disk id test
-	runNegativeListVolumeRequest(t, client, "-1")
```
why remove this test?

mauriciopoppe replied:
the API no longer receives strings or negative numbers because it's an unsigned int field now :), I think pre v1beta3 tests could be added again once I work on the e2e tests

jingxu97 commented on internal/os/volume/api.go:
```diff
@@ -11,12 +11,38 @@ import (
+	GetVolumeIDFromTargetPath(targetPath string) (string, error)
+	// WriteVolumeCache writes volume cache to disk
+	WriteVolumeCache(volumeID string) error
+}
```
match comments with above. but you can do it later once most review is done.

mauriciopoppe replied:
done

jingxu97 commented on internal/os/volume/api.go:
```diff
@@ -41,9 +67,9 @@ func getVolumeSize(volumeID string) (int64, error) {
-func (VolAPIImplementor) ListVolumesOnDisk(diskID string) (volumeIDs []string, err error) {
-	cmd := fmt.Sprintf("(Get-Disk -DeviceId %s |Get-Partition | Get-Volume).UniqueId", diskID)
+// ListVolumesOnDisk - returns back list of volumes(volumeIDs) in the disk.
+func (VolumeAPI) ListVolumesOnDisk(diskNumber uint32) (volumeIDs []string, err error) {
```
partition number is added as a second parameter

jingxu97 replied:
Get-Partition
   [[-DiskNumber] <UInt32[]>]
   [[-PartitionNumber] <UInt32[]>]
   [-CimSession <CimSession[]>]
   [-ThrottleLimit <Int32>]
   [-AsJob]
   [<CommonParameters>]

So I think you can use Get-Partition -DiskNumber %d -PartitionNumber %d

mauriciopoppe replied:
correct, I forgot to ask about using this parameter which is 0 by default, thanks for noticing this!

mauriciopoppe replied:
do you think that this is ok? The default value set by gRPC is [0 for unsigned int](https://developers.google.com/protocol-buffers/docs/proto3#default) so the command is like this:

```
// ListVolumesOnDisk - returns back list of volumes(volumeIDs) in a disk and a partition.
func (VolumeAPI) ListVolumesOnDisk(diskNumber uint32, partitionNumber uint32) (volumeIDs []string, err error) {
	// partitionNumber=0 means that the partitionNumber wasn't set so we list all the partitions
	var cmd string
	if partitionNumber == 0 {
		cmd = fmt.Sprintf("(Get-Disk -Number %d | Get-Partition | Get-Volume).UniqueId", diskNumber)
	} else {
		cmd = fmt.Sprintf("(Get-Disk -Number %d | Get-Partition -PartitionNumber %d | Get-Volume).UniqueId", diskNumber, partitionNumber)
	}
```

jingxu97 replied:
it looks fine.

jingxu97 commented on internal/server/volume/server.go:
```diff
@@ -221,38 +239,52 @@ func (s *Server) GetVolumeDiskNumber(context context.Context, request *internal.
 	minimumVersion := apiversion.NewVersionOrPanic("v1beta1")
 	if version.Compare(minimumVersion) < 0 {
-		return nil, fmt.Errorf("GetVolumeFromMount requires CSI-Proxy API version v1beta1 or greater")
+		return nil, fmt.Errorf("GetVolumeIDFromTargetPath requires CSI-Proxy API version v1beta1 or greater")
```
is GetVolumeIDFromTargetPath added to v1beta3?

mauriciopoppe replied:
this check was added in `GetVolumeIDFromMount`, I've moved it there instead

jingxu97 commented on internal/os/volume/api.go:
```diff
@@ -54,7 +80,7 @@ func (VolAPIImplementor) ListVolumesOnDisk(diskID string) (volumeIDs []string, e
 
 // FormatVolume - Format a volume with a pre specified filesystem (typically ntfs)
-func (VolAPIImplementor) FormatVolume(volumeID string) (err error) {
+func (VolumeAPI) FormatVolume(volumeID string) (err error) {
```
update comment, we only use ntfs

mauriciopoppe replied:
done

mauriciopoppe commented on internal/server/volume/server.go:
```diff
@@ -221,38 +239,52 @@ func (s *Server) GetVolumeDiskNumber(context context.Context, request *internal.
 
 func (s *Server) GetVolumeIDFromMount(context context.Context, request *internal.VolumeIDFromMountRequest, version apiversion.Version) (*internal.VolumeIDFromMountResponse, error) {
-	klog.V(4).Infof("calling GetVolumeFromMount with request %+v", request)
+	getVolumeIDFromTargetPathRequest := &internal.GetVolumeIDFromTargetPathRequest{
```
I think we could add this check however a newer client v1beta3 will be unable to reach this endpoint because it's not declared in its proto file, the same thing happens with an pre v1beta3 client that wants to reach some of the new v1beta3 endpoints, they would be unreachable too

jingxu97 commented on internal/server/volume/server.go:
```diff
@@ -181,6 +169,21 @@ func (s *Server) ResizeVolume(context context.Context, request *internal.ResizeV
+
+func (s *Server) GetVolumeStats(context context.Context, request *internal.GetVolumeStatsRequest, version apiversion.Version) (*internal.GetVolumeStatsResponse, error) {
 	klog.V(4).Infof("calling VolumeStats with request: %+v", request)
 	minimumVersion := apiversion.NewVersionOrPanic("v1beta1")
```
should change from v1beta1 to v1beta3 for GetVolumeStats?

mauriciopoppe replied:
thanks for noticing this, this belongs to `VolumeStats` and not `GetVolumeStats`

mauriciopoppe commented:
/unhold
Something that I haven't tested are the e2e tests, I could do this before we merge this or after all the other APIs have been updated @jingxu97.

jingxu97 commented:
> /unhold
> Something that I haven't tested are the e2e tests, I could do this before we merge this or after all the other APIs have been updated @jingxu97.

since this is the first big api change, maybe it worth to test it with e2e to discover any issues early?

mauriciopoppe commented:
/hold
@jingxu97 sure, I'll run the e2e tests first and add some pre v1beta3 tests too

mauriciopoppe commented:
/unhold

I've run the volume e2e tests in a GCE windows instance with Hyper-V enabled, these are the results:

```
PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> go clean -testcache
PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> go test -v .\integrationtests\ -run TestVolumeAPIs
=== RUN   TestVolumeAPIs
=== RUN   TestVolumeAPIs/E2E
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c mkdir C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\mount-75"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\disk-25.vhdx -SizeBytes 1073741824"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Mount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\disk-25.vhdx"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c (Get-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\disk-25.vhdx).DiskNumber"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Initialize-Disk -Number 14 -PartitionStyle GPT"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-Partition -DiskNumber 14 -UseMaximumSize"
    volume_test.go:407: VolumeId \\?\Volume{1abe379c-e431-4c8a-bb48-35bbe8ea30cd}\
    volume_test.go:253: Checking that the size is inside the bounds: 1052770304 < (actual) 1056894976 < 1073741824
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Resize-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\disk-25.vhdx -SizeBytes 2113789952"
    volume_test.go:446: Attempt to resize volume from sizeBytes=1056894976 to sizeBytes=1585342464
    volume_test.go:253: Checking that the size is inside the bounds: 1564370944 < (actual) 1585340416 < 1585342464
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Dismount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\disk-25.vhdx"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rm C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\disk-25.vhdx"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\mount-75"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-57.csi.io\\"
=== RUN   TestVolumeAPIs/NegativeDiskTests
=== RUN   TestVolumeAPIs/NegativeVolumeTests
=== RUN   TestVolumeAPIs/VolumeAPICompatibilityTests
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c mkdir C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\mount-66"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\disk-82.vhdx -SizeBytes 1073741824"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Mount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\disk-82.vhdx"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c (Get-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\disk-82.vhdx).DiskNumber"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Initialize-Disk -Number 14 -PartitionStyle GPT"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-Partition -DiskNumber 14 -UseMaximumSize"
    volume_test.go:253: Checking that the size is inside the bounds: 1052770304 < (actual) 1056894976 < 1073741824
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Dismount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\disk-82.vhdx"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rm C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\disk-82.vhdx"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\mount-66"
    volume_test.go:25: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-5.csi.io\\"
--- PASS: TestVolumeAPIs (104.49s)
    --- PASS: TestVolumeAPIs/E2E (57.27s)
    --- PASS: TestVolumeAPIs/NegativeDiskTests (0.00s)
    --- PASS: TestVolumeAPIs/NegativeVolumeTests (10.31s)
    --- PASS: TestVolumeAPIs/VolumeAPICompatibilityTests (36.91s)
PASS
ok      github.com/kubernetes-csi/csi-proxy/integrationtests    104.828s
```

mauriciopoppe commented:
@jingxu97 @ddebroy I've added tests for v1alpha1, v1beta1 and v1beta2, the output is:

```
PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> go test -v .\integrationtests\ -run TestVolumeAPIs

=== RUN   TestVolumeAPIs
=== RUN   TestVolumeAPIs/NegativeDiskTests
=== RUN   TestVolumeAPIs/NegativeVolumeTests
=== RUN   TestVolumeAPIs/v1alpha1Tests
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c mkdir C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\mount-6"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\disk-60.vhdx -SizeBytes 1073741824"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Mount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\disk-60.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c (Get-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\disk-60.vhdx).DiskNumber"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Initialize-Disk
-Number 17 -PartitionStyle GPT"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-Partition -D
iskNumber 17 -UseMaximumSize"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Resize-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\disk-60.vhdx -SizeBytes 2147483648"
    volume_v1alpha1_test.go:78: Attempt to resize volume from sizeBytes=1073741824 to sizeBytes=1610612736
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Dismount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\disk-60.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rm C:\\var\\lib\\kubelet\\plugins\\testplugin-68.csi.io\\disk-60.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\l
ib\\kubelet\\plugins\\testplugin-68.csi.io\\mount-6"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\l
ib\\kubelet\\plugins\\testplugin-68.csi.io\\"
=== RUN   TestVolumeAPIs/v1beta1Tests
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c mkdir C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\mount-52"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\disk-46.vhdx -SizeBytes 1073741824"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Mount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\disk-46.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c (Get-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\disk-46.vhdx).DiskNumber"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Initialize-Disk -Number 17 -PartitionStyle GPT"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-Partition -D
iskNumber 17 -UseMaximumSize"
    volume_test.go:129: Checking that the size is inside the bounds: 1052770304 < (actual) 1056894976 < 1073741824
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Resize-VHD -Path
 C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\disk-46.vhdx -SizeBytes 2147483648"
    volume_v1beta1_test.go:100: Attempt to resize volume from sizeBytes=1056894976 to sizeBytes=1585342464
    volume_test.go:129: Checking that the size is inside the bounds: 1564370944 < (actual) 1585340416 < 1585342464
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Dismount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\disk-46.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rm C:\\var\\lib\\kubelet\\plugins\\testplugin-18.csi.io\\disk-46.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\l
ib\\kubelet\\plugins\\testplugin-18.csi.io\\mount-52"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\l
ib\\kubelet\\plugins\\testplugin-18.csi.io\\"
=== RUN   TestVolumeAPIs/v1beta2Tests
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c mkdir C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\mount-70"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\disk-76.vhdx -SizeBytes 1073741824"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Mount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\disk-76.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c (Get-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\disk-76.vhdx).DiskNumber"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Initialize-Disk -Number 17 -PartitionStyle GPT"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-Partition -DiskNumber 17 -UseMaximumSize"
    volume_test.go:129: Checking that the size is inside the bounds: 1052770304 < (actual) 1056894976 < 1073741824
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Resize-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\disk-76.vhdx -SizeBytes 2147483648"
    volume_v1beta2_test.go:99: Attempt to resize volume from sizeBytes=1056894976 to sizeBytes=1585342464
    volume_test.go:129: Checking that the size is inside the bounds: 1564370944 < (actual) 1585340416 < 1585342464
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Dismount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\disk-76.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rm C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\disk-76.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\mount-70"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-63.csi.io\\"
=== RUN   TestVolumeAPIs/v1beta3Tests
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c mkdir C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\mount-65"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\disk-49.vhdx -SizeBytes 1073741824"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Mount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\disk-49.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c (Get-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\disk-49.vhdx).DiskNumber"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Initialize-Disk -Number 17 -PartitionStyle GPT"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c New-Partition -DiskNumber 17 -UseMaximumSize"
    volume_v1beta3_test.go:73: VolumeId \\?\Volume{5e5fba17-9a3c-4102-8abd-6048e294ca69}\
    volume_test.go:129: Checking that the size is inside the bounds: 1052770304 < (actual) 1056894976 < 1073741824
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Resize-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\disk-49.vhdx -SizeBytes 2147483648"
    volume_v1beta3_test.go:112: Attempt to resize volume from sizeBytes=1056894976 to sizeBytes=1585342464
    volume_test.go:129: Checking that the size is inside the bounds: 1564370944 < (actual) 1585340416 < 1585342464
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c Dismount-VHD -Path C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\disk-49.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rm C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\disk-49.vhdx"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\mount-65"
    utils.go:164: Executing command: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe /c rmdir C:\\var\\lib\\kubelet\\plugins\\testplugin-16.csi.io\\"
--- PASS: TestVolumeAPIs (206.18s)
    --- PASS: TestVolumeAPIs/NegativeDiskTests (0.00s)
    --- PASS: TestVolumeAPIs/NegativeVolumeTests (10.23s)
    --- PASS: TestVolumeAPIs/v1alpha1Tests (42.16s)
    --- PASS: TestVolumeAPIs/v1beta1Tests (49.65s)
    --- PASS: TestVolumeAPIs/v1beta2Tests (50.24s)
    --- PASS: TestVolumeAPIs/v1beta3Tests (53.90s)
PASS
ok      github.com/kubernetes-csi/csi-proxy/integrationtests    206.540s
```

jingxu97 commented on client/api/volume/v1beta3/api.proto:
```diff
@@ -5,128 +5,129 @@ package v1beta3;
 service Volume {
-    // ListVolumesOnDisk returns the volume IDs (in \\.\Volume{GUID} format) for
-    // all volumes on a Disk device
+    // ListVolumesOnDisk returns the volume IDs (in \\.\Volume{GUID} format) for all volumes on a Disk device.
```
maybe ListVolumesOnDisk returns the volume IDs (in \\.\Volume{GUID} format) for all volumes from a given disk number and partition number (optional)

mauriciopoppe replied:
fixed

jingxu97 commented on integrationtests/volume_v1beta2_test.go:
```diff
@@ -0,0 +1,131 @@
+	}
+	// For a volume formatted with 1GB it should be around 1GB, in practice it was 1056947712 bytes or 0.9844GB
+	// let's compare with a range of - 20MB
+	if !sizeIsAround(t, volumeStatsResponse.VolumeSize, vhd.InitialSize) {
```
how it was handled before your change?

mauriciopoppe replied:
this wasn't tested before, I added this part

jingxu97 commented on integrationtests/volume_test.go:
```diff
@@ -196,14 +111,25 @@ func negativeVolumeTests(t *testing.T) {
+// (considers the fact that some bytes were lost)
+func sizeIsAround(t *testing.T, actualSize, expectedSize int64) bool {
+	// An upper bound on the number of bytes that are lost when creating or resizing a partition
+	var volumeSizeBytesLoss int64 = (20 * 1024 * 1024)
```
how this volumeSizeBytesLoss is obtained? Is there some doc related to this?

mauriciopoppe replied:
I got this value from looking at the GetStats output after you format a partition, if you format it to 1GB you'r getting actually ~0.9844GB, probably @ddebroy might help us understanding where are those bytes going :)

Anyways if you do 1GB - 20MB that's around 0.98GB which I thought it was a good lower bound

jingxu97 commented:
/lgtm

mauriciopoppe commented on internal/os/volume/api.go:
```diff
@@ -177,36 +213,36 @@ func (VolAPIImplementor) VolumeStats(volumeID string) (int64, int64, error) {
 
 	volumeUsedSize := volumeSize - volumeSizeRemaining
-	return volumeSizeRemaining, volumeUsedSize, nil
+	return volumeSize, volumeUsedSize, nil
```
This issue was raised during one of our meetings:

```
	totalBytes, usedBytes, err := s.hostAPI.GetVolumeStats(volumeID)

	volumeSize = getVolume["Size"]
	volumeSizeRemaining = getVolume["SizeRemaining"]
	volumeUsedSize := volumeSize - volumeSizeRemaining
	return volumeSizeRemaining, volumeUsedSize, nil

      // proposed
	return volumeSize, volumeUsedSize, nil
```

> Have we ever attempted to do a resize? I tried running a resize from 1GB to 2GB and I get:

```
PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> Get-Volume -UniqueId "\\?\Volume{1b6810f3-20a3-4a2f-bd7b-3b36f05265d4}\" | Get-partition


   DiskPath: \\?\scsi#disk&ven_msft&prod_virtual_disk#2&1f4adffe&0&000006#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}

PartitionNumber  DriveLetter Offset                                        Size Type
---------------  ----------- ------                                        ---- ----
2                            16777216                                1007.94 MB Basic


PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> Get-Volume -UniqueId "\\?\Volume{1b6810f3-20a3-4a2f-bd7b-3b36f05265d4}\" | Get-partition | Resize-Partition -Size 2147483648
Resize-Partition : Not enough available capacity
Activity ID: {229be40a-070d-41ff-b027-d71015bacf61}
At line:1 char:92
+ ... b-3b36f05265d4}\" | Get-partition | Resize-Partition -Size 2147483648
+                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (StorageWMI:ROOT/Microsoft/.../MSFT_Partition) [Resize-Partition], CimException
    + FullyQualified
```
